### PR TITLE
[WIP] Secure browser example

### PR DESCRIPTION
**Pull request by blink1073**
I followed the instructions as best as I can, but `@web.authenticated` requires the user to be actually logged in for anything but `GET` and `HEAD` requests, and I can't figure out how to actually perform the login.

"Failed to load resource: the server responded with a status of 403 (Forbidden)"

http://jupyter-notebook.readthedocs.io/en/latest/public_server.html

@minrk, lifeline?

cf #205

### CODE REVIEW

File: examples/browser-secure/main.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        return self.render("index.html", static=self.static_url,
+                           base_url=self.base_url)
+
+
+def main():
+    # Start a notebook server with cross-origin access.
+    url = "http://localhost:%s" % PORT
+
+    nb_command = [sys.executable, '-m', 'notebook', '--no-browser', '--debug',
+                  '--NotebookApp.password="%s"' % passwd('password'),
```
Comment (blink1073) [inline examples/browser-secure/main.py]: Relevant change is here.

File: examples/browser-secure/index.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+
+require(['jquery', 'jupyter-js-services'], function ($, services) {
+  'use strict';
+  var startNewKernel = services.startNewKernel;
+
+  var kernelOptions = {
+    name: 'python',
+    ajaxSettings: {
```
Comment (blink1073) [inline examples/browser-secure/index.js]: Other relevant change.

**Comment by minrk** [comment]
For now, a cookie has to be requested. It's uses like this that prompt the desire to add API Token-based authentication (like JupyterHub, Kernel Gateway), which is very simple, but not yet implemented.

To do the cookie auth, you have to POST the login form, and then use those cookies in future requests. With requests in Python:

```python
import requests
base_url = 'http://127.0.0.1:8888'

# make login request to get cookie
login = requests.post(base_url + '/login', data={
        'password': password
    }, allow_redirects=False)
login.raise_for_status()

# pass cookies=login.cookies to authorize future requests
r = requests.get(base_url + '/api/kernels/', cookies=login.cookies)
kernels = r.json()
```

**Comment by minrk** [comment]
In the case of a webapp like this one, you could check for 403, then redirect to the notebook's login page and redirect back here. I can't remember of cookies are sent with cross-origin requests, though.

**Comment by blink1073** [comment]
I futzed with this a bit more.  Sending data on the POST results in an OPTIONS request, which is not supported by the `login/` handler.  I don't think this particular example will work, I'm at a loss.

**Comment by minrk** [comment]
I'll take a stab at it with the redirect. I think browsers probably don't let you do cross-origin login stuff like you would in a Python/nodejs script.

**Comment by minrk** [comment]
I opened https://github.com/blink1073/jupyter-js-services/pull/8 to use the notebook's own login page, and then it works.

**Comment by blink1073** [comment]
I'll fix #208 before finishing this.

**Comment by minrk** [comment]
Cool, thanks!

**Comment by blink1073** [comment]
This is what we get back from the server now:

screen shot 2016-07-29 at 6 19 53 am

**Comment by blink1073** [comment]
Closing as stale.

**Pull request closed without merging**